### PR TITLE
SWATCH-1810 Do not send usage to AWS when too old

### PR DESCRIPTION
**Pull request by mstead**
Jira issue: [SWATCH-1810](https://issues.redhat.com/browse/SWATCH-1810)

## Description
The AWS Marketplace API will only accept usage in which
the timestamp must be on or after the START_OF_CURRENT_HOUR - 6h.

In order to avoid sending requests to the AWS API that we know will
fail, the AWS producer should check that the incoming usage is within
this window before sending the usage to AWS.

If the usage is outside of this window, the usage should not
be sent to AWS, a WARNING message should be logged, and an
'ignored' micrometer counter should be incremented.

An AWS_MARKETPLACE_USAGE_WINDOW configuration property was added to
allow us to adjust our pre-send window quickly if we ever need to.
The default value for this property is 6h.

## Example
Let's say that it is currently 2023-10-26T13:30:00Z. Any message that has a timestamp that is on or after 2023-10-26T07:00:00Z will be accepted. Therefore, a message sent with a timestamp of 2023-10-26T07:00:00Z  would be accepted up until the current time is 2023-10-26T14:00:00Z.

### Setup
1. Refresh the database
```
dropdb -U rhsm-subscriptions rhsm-subscriptions && createdb -U rhsm-subscriptions rhsm-subscriptions
```
2. Launch the subscriptions API, running on a different port than the AWS producer will be.
```
DEV_MODE=true SERVER_PORT=8101 ./gradlew :bootRun
```
3. Load a subscription into the DB.
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "insert into subscription (sku, org_id, subscription_id, quantity, start_date, end_date, billing_provider_id, subscription_number, billing_provider, billing_account_id) values ('BASILISK', 'org123', '111111112', 1, '2022-10-10T04:00:00Z', '2024-10-10T04:59:59.999Z', 'basilisk;customer1;awsSeller1', '222222222', 'aws', '123456');"

```

1. Launch the AWS billing producer, configuring it to point at the internal subscription API above.
```
SWATCH_INTERNAL_SUBSCRIPTION_ENDPOINT=http://localhost:8101/api/rhsm-subscriptions/v1  ENABLE_AWS_DRY_RUN=true ./gradlew swatch-producer-aws:quarkusDev
```

### Steps
**In the following examples, it is currently 18:40 UTC. Be sure to update the date/time of the examples below.**

1. Check that a WARNING is logged when the incoming usage is outside of the usage window (START_OF_CURRENT_HOUR - 6h).
```
curl -X 'POST'   'http://localhost:8000/api/swatch-producer-aws/internal/aws/billable_usage'   -H 'accept: */*'   -H 'Content-Type: application/json'   -d '{
  "account_number": "account123",
  "org_id": "org123",
  "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "billing_provider": "aws",
  "billing_account_id": "123456",
  "product_id": "BASILISK",
  "sla": "Premium",
  "usage": "Production",
  "uom": "Instance-hours",
  "value": 10,
  "billing_factor": 1,
  "vendor_product_code": "aws123",
  "snapshot_date": "2023-10-27T11:59:59.999999Z"
}'
```

Expected log output:
```log
2023-10-27 15:42:20,073 WARN  [com.red.swa.aws.pro.BillableUsageProcessor] (vert.x-worker-thread-1) SWATCHAWS1007: Usage timestamp will not be accepted by AWS: Unable to send usage since it is outside of the AWS processing window orgId=org123 tallySnapshotId=3fa85f64-5717-4562-b3fc-2c963f66afa6 productId=BASILISK snapshotDate=2023-10-27T11:59:59.999999Z rhSubscriptionId=111111112 awsCustomerId=customer1 awsProductCode=basilisk subscriptionStartDate=2022-10-10T04:00Z value=10
```

2. Send usage that JUST fits within the window and it should succeed.
```
curl -X 'POST'   'http://localhost:8000/api/swatch-producer-aws/internal/aws/billable_usage'   -H 'accept: */*'   -H 'Content-Type: application/json'   -d '{
  "account_number": "account123",
  "org_id": "org123",
  "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "billing_provider": "aws",
  "billing_account_id": "123456",
  "snapshot_date": "2023-10-27T12:00:00Z",
  "product_id": "BASILISK",
  "sla": "Premium",
  "usage": "Production",
  "uom": "Instance-hours",
  "value": 10,
  "billing_factor": 1,
  "vendor_product_code": "aws123"
}'
```
Expected log output:
```log
2023-10-27 15:42:30,073 INFO  [com.red.swa.aws.pro.BillableUsageProcessor] (vert.x-worker-thread-1) [DRY RUN] Sending usage request to AWS: BatchMeterUsageRequest(UsageRecords=[UsageRecord(Timestamp=2023-10-27T12:00:00Z, CustomerIdentifier=customer1, Dimension=cluster_hour, Quantity=10)], ProductCode=basilisk), organization=org123, product_id=BASILISK

```

3. Send usage that fits within the window and it should succeed.
```
curl -X 'POST'   'http://localhost:8000/api/swatch-producer-aws/internal/aws/billable_usage'   -H 'accept: */*'   -H 'Content-Type: application/json'   -d '{
  "account_number": "account123",
  "org_id": "org123",
  "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "billing_provider": "aws",
  "billing_account_id": "123456",
  "snapshot_date": "2023-10-27T15:00:00Z",
  "product_id": "BASILISK",
  "sla": "Premium",
  "usage": "Production",
  "uom": "Instance-hours",
  "value": 10,
  "billing_factor": 1,
  "vendor_product_code": "aws123"
}'
```
Expected log output:
```log
2023-10-27 15:42:40,208 INFO  [com.red.swa.aws.pro.BillableUsageProcessor] (vert.x-worker-thread-1) [DRY RUN] Sending usage request to AWS: BatchMeterUsageRequest(UsageRecords=[UsageRecord(Timestamp=2023-10-27T15:00:00Z, CustomerIdentifier=customer1, Dimension=cluster_hour, Quantity=10)], ProductCode=basilisk), organization=org123, product_id=BASILISK

```

### CODE REVIEW

File: swatch-producer-aws/src/main/java/com/redhat/swatch/aws/exception/UsageTooOldException.java
Comment (Sgitario) [inline swatch-producer-aws/src/main/java/com/redhat/swatch/aws/exception/UsageTooOldException.java:25]: I would add a javadoc with a link to https://docs.aws.amazon.com/marketplacemetering/latest/APIReference/API_UsageRecord.html and say that the "Timestamp" property is what matters here. 

Moreover, I would rename this class "UsageTooOldException" to "UsageTimestampOutOfBoundsException" or something like this. Yet, if you would like to keep using "UsageTooOldException", I'm ok too.

**Review by Sgitario — COMMENTED** [review]
LGTM. Only one comment.

File: swatch-producer-aws/src/main/java/com/redhat/swatch/aws/exception/UsageTooOldException.java
Comment (mstead) [reply swatch-producer-aws/src/main/java/com/redhat/swatch/aws/exception/UsageTooOldException.java:25]: Solid suggestions, I'll make the changes. Thanks!

Comment (mstead) [reply swatch-producer-aws/src/main/java/com/redhat/swatch/aws/exception/UsageTooOldException.java]: Done.

**Review by Sgitario — APPROVED** [review]
LGTM

**Comment by mstead** [comment]
**NOTE:** QE is actively working on getting this reviewed.

**Comment by Aurobinda55** [comment]
There are some test setup issues on my EE deployment and I am still working on it to fix them and verify this. Will update tomorrow.

**Review by san7ket — APPROVED** [review]
EE lacks Automation, Tested scenario manually

**Comment by mstead** [comment]
Just realized that I forgot to add the config property to the cloudapp file. I'll make that change shortly.

**Comment by kartikshahc** [comment]
Do not merge!

**Comment by mstead** [comment]
The 1h AWS window that we have been talking about for quite some time, may not be correct. Multiple places in the AWS API docs suggest that the window is either 1h or 6h. 

Before merging this PR, we need to determine what the ACTUAL window is and update the config value appropriately.

**Comment by mstead** [comment]
> The 1h AWS window that we have been talking about for quite some time, may not be correct. Multiple places in the AWS API docs suggest that the window is either 1h or 6h.
> 
> Before merging this PR, we need to determine what the ACTUAL window is and update the config value appropriately.

After playing around with the API, the window is 6h as documented, however, it doesn't necessarily represent a message age of 6h.

A message timestamp must be **on or after the START_OF_CURRENT_HOUR - 6h**

For example, let's say that it is currently `2023-10-26T13:30:00Z`. Any message that has a timestamp that is on or after `2023-10-26T07:00:00Z` will be accepted. Therefore, a message sent with a timestamp of `2023-10-26T07:00:00Z ` would be accepted up until the current time is `2023-10-26T14:00:00Z`.

Here's some more examples, using the current time of 2023-10-26T13:30:00Z. Consider message with the following timestamp:
* 2023-10-26T13:00:00Z == VALID
* 2023-10-26T07:00:00Z == VALID
* 2023-10-26T06:59:59.999999Z == INVALID
* 2023-10-26T05:30:00Z == INVALID

I will update this PR to apply the same sort of formula before applying the configured duration.

**Comment by mstead** [comment]
@Sgitario @Aurobinda55 I've made the required changes to this PR. When you have time, please give it one more check.

File: swatch-producer-aws/src/main/java/com/redhat/swatch/aws/exception/UsageTimestampOutOfBoundsException.java
Comment (Sgitario) [inline swatch-producer-aws/src/main/java/com/redhat/swatch/aws/exception/UsageTimestampOutOfBoundsException.java:27]:
```suggestion
 * timestamp range.
```

**Review by Sgitario — APPROVED** [review]
LGTM. 

So, the cause of getting these exceptions in the previous deployment was because the events were 6h old?

**Comment by mstead** [comment]
> So, the cause of getting these exceptions in the previous deployment was because the events were 6h old?

Yes. Re-tally was being triggered, which sent a large number of BillableUsage messages to the BillableUsageProcessor that were outside of the usage window.

**Review by san7ket — APPROVED** [review]
This proved a bit difficult to test with our automation and the recent changes that went into the timestamp/record_date of events generation. 

In the end, had to resort to check with the internal billing_usage api endpoint.
Looks, like all the events are processed within the window and error is successfully produced for events outside of 6h window. 

- Sync offering

- Add Sub to account 

```
internal_api_add_sku("MCT4121HR", app.user.identity.org_id, billing_provider="aws", billing_account_id=102, start_day="2023-10-21")
2023-10-31 10:30:53.975 [    INFO] [root] POST request response url: http://swatch-subscription-sync-service:8000/api/rhsm-subscriptions/v1/internal/subscriptions?reconcileCapacity=true, status_code: 200
2023-10-31 10:30:53.975 [    INFO] [root] Response: {
    "detail": "Success"
}
Out[30]: 
[{'id': '000149',
  'subscriptionNumber': '8838431',
  'subscriptionProducts': [{'sku': 'MCT4121HR'}],
  'webCustomerId': '14541308',
  'quantity': 1,
  'oracleAccountNumber': '',
  'effectiveStartDate': 1697847432103.0,
  'effectiveEndDate': 1730247432103.0,
  'externalReferences': {'awsMarketplace': {'accountID': None,
    'subscriptionID': '100',
    'customerID': 102,
    'productCode': 'testProductCode',
    'sellerAccount': '1234567',
    'customerAccountID': 102}}}]

```

- Send post to the internal billing usage api 
```
curl -X 'POST'   'http://localhost:8000/api/swatch-producer-aws/internal/aws/billable_usage'   -H 'accept: */*'   -H 'Content-Type: application/json' -H 'x-rh-swatch-psk: placeholder'  -d '{
  "org_id": "14541308",
  "id": "fde18074-a9d7-4fd3-99f9-85d139480be3",
  "billing_provider": "aws",
  "billing_account_id": "102",
  "snapshot_date": "2023-10-31T03:10:50Z",
  "product_id": "rhods",
  "sla": "Premium",
  "usage": "Production",
  "uom": "Cores",
  "value": 93,
  "billing_factor": 1
}'
```

Responses:
```
hat.swatch.aws.processors.BillableUsageProducer","level":"INFO","message":"Queued up a BillableUsage object to the billable-usage topic","threadName":"executor-thread-12","threadId":44,"mdc":{"s}
{"timestamp":"2023-10-31T10:33:46.547352039Z","sequence":225,"loggerClassName":"org.slf4j.impl.Slf4jLogger","loggerName":"com.redhat.swatch.aws.processors.BillableUsageProcessor","level":"DEBUG","message":"Picked up billable usage message class BillableUsage {\n orgId: 14541308\n id: fde18074-a9d7-4fd3-99f9-85d139480be3\n billingProvider: aws\n billingAccountId: 102\n snapshotDate: 2023-10-31T03:10:50Z\n productId: rhods\n sla: Premium\n usage: Production\n uom: Cores\n value: 93\n billingFactor: 1\n vendorProductCode: null\n hardwareMeasurementType: null\n} to process","threadName":"executor-thread-12","threadId":44,"mdc":{"spanId":"80c7c8d62cfcca6c","traceId":"e70b2da1ddcafec901321a61f50bd5c3","sampled":"true"},"ndc":"","hostName":"swatch-producer-aws-service-7b97f8d9bb-mdbp7","processName":"quarkus-run.jar","processId":1}
{"timestamp":"2023-10-31T10:33:46.559961136Z","sequence":226,"loggerClassName":"org.slf4j.impl.Slf4jLogger","loggerName":"com.redhat.swatch.aws.processors.BillableUsageProcessor","level":"WARN","message":"SWATCHAWS1007: Usage timestamp will not be accepted by AWS: Unable to send usage since it is outside of the AWS processing window orgId=14541308 tallySnapshotId=fde18074-a9d7-4fd3-99f9-85d139480be3 productId=rhods snapshotDate=2023-10-31T03:10:50Z rhSubscriptionId=149 awsCustomerId=102 awsProductCode=testProductCode subscriptionStartDate=2023-10-21T00:17:12.103Z value=93","threadName":"executor-thread-12","threadId":44,"mdc":{"org_id":"14541308","spanId":"80c7c8d62cfcca6c","traceId":"e70b2da1ddcafec901321a61f50bd5c3","sampl
```

**Pull request merged**
